src/Npgsql/TaskExtensions.cs

```diff
@@ -8,122 +9,127 @@ namespace Npgsql
+            if (cancellationToken.CanBeCanceled)
             {
-                var combinedCancellationToken = GetCombinedCancellationToken(ref combinedCts, timeout, cancellationToken);
-                await func(combinedCancellationToken);
+                var tcs = new TaskCompletionSource<bool>();
+                registration = cancellationToken.Register(s => ((TaskCompletionSource<bool>)s!).TrySetResult(true), tcs);
+                cancellationTask = tcs.Task;
+                tasks.Value.Add(cancellationTask);
             }
-            finally
+
+            Task? delayTask = default;
+            CancellationTokenSource? delayCts = default;
+            if (timeout.IsSet)
             {
-                combinedCts?.Dispose();
+                var timeLeft = timeout.CheckAndGetTimeLeft();
+                delayCts = new CancellationTokenSource();
+                delayTask = Task.Delay(timeLeft, delayCts.Token);
+                tasks.Value.Add(delayTask);
             }
-        }
 
-        internal static async Task<TResult> ExecuteWithTimeout<TResult>(Func<CancellationToken, Task<TResult>> func, NpgsqlTimeout timeout, CancellationToken cancellationToken)
-        {
-            CancellationTokenSource? combinedCts = null;
             try
             {
-                var combinedCancellationToken = GetCombinedCancellationToken(ref combinedCts, timeout, cancellationToken);
-                return await func(combinedCancellationToken);
+                if (tasks.IsValueCreated)
+                {
+                    tasks.Value.Add(task);
+                    var result = await Task.WhenAny(tasks.Value);
+                    if (result == cancellationTask)
+                    {
+                        task = Task.FromCanceled(cancellationToken);
+                    }
+                    else if (result == delayTask)
+                    {
+                        task = Task.FromException(new TimeoutException());
+                    }
+                }
+                await task;
             }
             finally
             {
-                combinedCts?.Dispose();
+                delayCts?.Cancel();
+                delayCts?.Dispose();
+                registration.Dispose();
             }
         }
-
-        static CancellationToken GetCombinedCancellationToken(ref CancellationTokenSource? combinedCts, NpgsqlTimeout timeout, CancellationToken cancellationToken)
-        {
-            var finalCt = cancellationToken;
-
-            if (timeout.IsSet)
-            {
-                combinedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
-                combinedCts.CancelAfter((int)timeout.CheckAndGetTimeLeft().TotalMilliseconds);
-                finalCt = combinedCts.Token;
-            }
-
-            return finalCt;
-        }
+        
+#if NET6_0_OR_GREATER
+        static Task WithCancellationAndTimeout(Task task, NpgsqlTimeout timeout, CancellationToken cancellationToken)
```

Comment:
@roji FYI I'll move this line out of if/else/endif in my next commit to avoid duplication
```
        static Task WithCancellationAndTimeout(Task task, NpgsqlTimeout timeout, CancellationToken cancellationToken) =>
#if NET6_0_OR_GREATER
            WithCancellationAndTimeoutImpl(task, timeout, cancellationToken);
#else
            WithCancellationAndTimeoutImplLegacy(task, timeout, cancellationToken);
#endif
```